modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/IncrementalSnapshotVerificationTask.java

```diff
@@ -0,0 +1,421 @@
+            this.snpPath = snpPath;
+            this.incIdx = incIdx;
+            this.consId = consId;
+        }
+
+        /**
+         * @return Map containing calculated transactions hash for every remote node in the cluster.
+         */
+        @Override public IncrementalSnapshotVerificationTaskResult execute() throws IgniteException {
+            try {
+                if (log.isInfoEnabled()) {
+                    log.info("Verify incremental snapshot procedure has been initiated " +
+                        "[snpName=" + snpName + ", incrementIndex=" + incIdx + ", consId=" + consId + ']');
+                }
+
+                if (incIdx <= 0)
+                    return new IncrementalSnapshotVerificationTaskResult();
+
+                BaselineTopology blt = ignite.context().state().clusterState().baselineTopology();
+
+                checkBaseline(blt);
+
+                Map<Integer, StoredCacheData> txCaches = readTxCachesData();
+
+                IncrementalSnapshotProcessor proc = new IncrementalSnapshotProcessor(
+                    ignite.context().cache().context(), snpName, snpPath, incIdx, txCaches.keySet()
+                ) {
+                    @Override void totalWalSegments(int segCnt) {
+                        // No-op.
+                    }
+
+                    @Override void processedWalSegments(int segCnt) {
+                        // No-op.
+                    }
+
+                    @Override void initWalEntries(LongAdder entriesCnt) {
+                        // No-op.
+                    }
+                };
+
+                short locShortId = blt.consistentIdMapping().get(consId);
+
+                Set<GridCacheVersion> activeDhtTxs = new HashSet<>();
+                Map<GridCacheVersion, Set<Short>> txPrimPartNodes = new HashMap<>();
+                Map<Short, Integer> nodesTxHash = new HashMap<>();
+
+                Set<GridCacheVersion> partialCommittedTxs = new HashSet<>();
+                Map<PartitionKeyV2, PartitionHashHolder> partMap = new HashMap<>();
+                List<Exception> exceptions = new ArrayList<>();
+
+                BiConsumer<GridCacheVersion, Set<Short>> calcTransactionHash = (xid, partNodes) -> {
+                    for (short shortId: partNodes) {
+                        if (shortId != locShortId)
+                            nodesTxHash.compute(shortId, (id, hash) -> xid.hashCode() + (hash == null ? 0 : hash));
+                    }
+                };
+
+                proc.process(dataEntry -> {
+                    if (dataEntry.op() == GridCacheOperation.READ)
+                        return;
+
+                    if (!activeDhtTxs.contains(dataEntry.writeVersion()))
+                        partialCommittedTxs.add(dataEntry.nearXidVersion());
+
+                    StoredCacheData cacheData = txCaches.get(dataEntry.cacheId());
+
+                    int grpId = CU.cacheGroupId(cacheData.config().getName(), cacheData.config().getGroupName());
```

Comment:
Let's cache cacheId -> grpId mapping somehow.
No need to recalculate value on each data entry.